packages/flutter/lib/src/services/text_input.dart

```diff
@@ -16,47 +16,118 @@ export 'dart:ui' show TextAffinity;
   ///
   /// Requests the default platform keyboard.
-  text,
+  static const TextInputType text = const TextInputType._(0);
 
   /// Optimize for multi-line textual information.
   ///
   /// Requests the default platform keyboard, but accepts newlines when the
   /// enter key is pressed. This is the input type used for all multi-line text
   /// fields.
-  multiline,
+  static const TextInputType multiline = const TextInputType._(1);
 
   /// Optimize for numerical information.
   ///
-  /// Requests a keyboard with ready access to the decimal point and number
-  /// keys.
-  number,
+  /// Requests a default keyboard with ready access to the number keys.
+  /// Additional settings, such as decimal point and/or positive/negative
+  /// signs, can be requested by using the [numberWithOptions] constructor.
+  static const TextInputType number = const TextInputType.numberWithOptions();
 
   /// Optimize for telephone numbers.
   ///
   /// Requests a keyboard with ready access to the number keys, "*", and "#".
-  phone,
+  static const TextInputType phone = const TextInputType._(3);
 
   /// Optimize for date and time information.
   ///
   /// On iOS, requests the default keyboard.
   ///
   /// On Android, requests a keyboard with ready access to the number keys,
   /// ":", and "-".
-  datetime,
+  static const TextInputType datetime = const TextInputType._(4);
 
   /// Optimize for email addresses.
   ///
   /// Requests a keyboard with ready access to the "@" and "." keys.
-  emailAddress,
+  static const TextInputType emailAddress = const TextInputType._(5);
 
   /// Optimize for URLs.
   ///
   /// Requests a keyboard with ready access to the "/" and "." keys.
-  url,
+  static const TextInputType url = const TextInputType._(6);
+
+  static const List<TextInputType> values = const <TextInputType>[
+    text, multiline, number, phone, datetime, emailAddress, url,
+  ];
+
+  static const List<String> _names = const <String>[
+    'text', 'multiline', 'number', 'phone', 'datetime', 'emailAddress', 'url',
+  ];
+
+  /// Enum value name, this is what enum.toString() would normally return.
+  String get _name => 'TextInputType.${_names[type]}';
+
+  /// Returns a representation of this object as a JSON object.
+  Map<String, dynamic> toJSON() {
+    return <String, dynamic>{
+      'type': _name,
+      'signed': signed,
+      'decimal': decimal,
+    };
+  }
+
+  @override
+  String toString() => '$runtimeType('
+      'type: \u2524$_name\u251C, '
+      'signed: $signed, '
+      'decimal: $decimal)';
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other is! TextInputType)
+      return false;
+    final TextInputType typedOther = other;
+    return typedOther.type == type
+        && typedOther.signed == signed
+        && typedOther.decimal == decimal;
+  }
+
+  @override
+  int get hashCode => hashValues(
+      type.hashCode,
+      signed.hashCode,
+      decimal.hashCode
```

Comment:
indent by 2 rather than 4.

This is a case where `=>` looks ok, though we usually would either do it all on one line, or use a block with return.